app/services/api/public/node/filter.rb

```diff
@@ -0,0 +1,125 @@
+module Api
+  module Public
+    module Node
+      class Filter
+        attr_reader :data
+
+        def initialize(params = {})
+          initialize_params(params)
+        end
+
+        def call
+          initialize_query
+
+          apply_filters
+
+          @query.first
+        end
+
+        private
+
+        def initialize_params(params)
+          @node_id = params[:id]
+          @year = params[:year]
+        end
+
+        def initialize_query
+          @query = Api::V3::Readonly::FlowNode.
+            select(*select_clause).
+            joins('INNER JOIN nodes_mv ON nodes_mv.id = flow_nodes_mv.node_id').
+            joins('INNER JOIN contexts_mv ON contexts_mv.id = flow_nodes_mv.context_id').
+            group(*group_clause).
+            order('flow_nodes_mv.node_id')
+        end
+
+        def select_clause
+          [
+            'flow_nodes_mv.node_id',
+            'nodes_mv.name',
+            'nodes_mv.node_type',
+            'nodes_mv.geo_id',
+            'JSON_AGG(' \
+              'DISTINCT JSONB_BUILD_OBJECT(' \
+                '\'country\', contexts_mv.iso2, ' \
+                '\'commodity\', contexts_mv.commodity_name, ' \
+                '\'years\', nodes_mv.years' \
+              ')' \
+            ') AS availability',
+            'JSON_AGG(' \
+              'DISTINCT JSONB_BUILD_OBJECT(' \
+                '\'country\', contexts_mv.iso2, ' \
+                '\'commodity\', contexts_mv.commodity_name, ' \
+                "'node_quants', #{select_node_clause('quants')}, " \
+                "'node_quals', #{select_node_clause('quals')}, " \
+                "'node_inds', #{select_node_clause('inds')}" \
+              ')' \
+            ') AS node_indicators',
+            'JSON_AGG(' \
+              'DISTINCT JSONB_BUILD_OBJECT(' \
+                '\'country\', contexts_mv.iso2, ' \
+                '\'commodity\', contexts_mv.commodity_name, ' \
+                '\'flow_id\', flow_nodes_mv.flow_id, ' \
+                "'flow_quants', #{select_flow_clause('quants')}, " \
```

Comment:
I'd get rid of the `flow_quants`, `flow_quals` and `flow_inds` keys (and the same for `node_quants` etc), because public API users will be confused about what a qual, quant and ind is. Currently in the public attributes endpoint  `/api/public/attributes` we're not exposing this internal type, because probably there's no advantage to external users to know this. So in interest of avoiding having to respond to questions "what is an ind", I'd just mix all the values together. Currently we have the structure:
```
flowIndicators: [
  {
    country:
    flowId:
    commodity:
    flowInds: [{name: ..., year: ..., value: ...}, ...]
    flowQuals: ...
    flowQuants:
  }
]
```

maybe we could simplify:

```
flowAttributes: [
{
  country:
  commodity:
  values: [{name: quant name, value: ..., year: ...}, {name: ind name, value: ..., year: ...}]
}
]
```

Also, I think since we expose attributes ids in the `/api/public/attributes` endpoint, I guess instead of attribute names here we could use the attributes ids, which would make the response shorter. Not sure if it would make the query easier, so this is just a suggestion.

Reply:
I have included all nodes on the same `values` property